test/wait-task/pkg/reconciler/reconciler.go

```diff
@@ -67,27 +74,67 @@ func (c *Reconciler) ReconcileKind(ctx context.Context, r *v1alpha1.Run) kreconc
 		return nil
 	}
 
-	dur, err := time.ParseDuration(expr.Value.StringVal)
-	if err != nil {
-		r.Status.MarkRunFailed("InvalidDuration", "The duration param was invalid: %v", err)
+	// Skip if the Run is cancelled.
+	if r.IsCancelled() {
+		logger.Infof("The Custom Task Run %v has been cancelled", r.GetName())
+		// Check if the Run has been cancelled by its Parent PipelineRun
+		if r.Spec.Status == v1alpha1.RunSpecStatusCancelled && r.Spec.StatusMessage == v1alpha1.RunCancelledByPipelineMsg {
+			now := metav1.Now()
+			r.Status.CompletionTime = &now
+			r.Status.MarkRunFailed("ParentPipelineRunTimedOut", WaitTaskCancelledByPipelineRunTimeoutMsg)
+		}
 		return nil
 	}
 
-	if r.Status.StartTime == nil {
-		now := metav1.Now()
-		r.Status.StartTime = &now
+	if !r.HasStarted() {
+		logger.Info("Run hasn't stated, start it")
+		r.Status.InitializeConditions()
 		r.Status.MarkRunRunning("Running", "Waiting for duration to elapse")
 	}
 
+	dur, err := time.ParseDuration(expr.Value.StringVal)
+	if err != nil {
+		r.Status.MarkRunFailed("InvalidDuration", "The duration param was invalid: %v", err)
+		return nil
+	}
 	done := r.Status.StartTime.Time.Add(dur)
 
-	if time.Now().After(done) {
+	to := time.Minute * config.DefaultTimeoutMinutes
+	if r.GetTimeout() != 0 {
+		to, err = time.ParseDuration(r.GetTimeout().String())
+		if err != nil {
+			r.Status.MarkRunFailed("InvalidTimeOut", "Spec.Timeout was invalid: %v", err)
+			return nil
+		}
+	}
+	timeout := r.Status.StartTime.Time.Add(to)
+
+	// Custom Task is running and not timed out
+	if r.Status.StartTime != nil && time.Now().Before(done) && time.Now().Before(timeout) {
```

Comment:
instead of using time.Now, it might make sense to use Clock like the other reconcilers do-- this makes it easy to provide a fake clock for unit testing.